src/solvers/td3.ml

```diff
@@ -234,10 +244,45 @@ module WP =
         )
       and eq x get set =
         if tracing then trace "sol2" "eq %a\n" S.Var.pretty_trace x;
-        eval_rhs_event x;
+        
         match S.system x with
         | None -> S.Dom.bot ()
-        | Some f -> f get set
+        | Some f ->
+          if skip_unchanged_rhs then
+            let all_deps_unchanged =
+              match HM.find_option dep_vals x with
+              | None -> None
+              | Some (oldv, deps) ->
```

Comment:
Is it necessary to store `oldv` like this? Might as well take it directly from `rho` if dependencies are unchanged.
At least that worked for #364 where no values where stored for aborting at all. In `caching-constrsys` the only reason I stored a separate `prev_val` was because the functor had no access to solver's `rho`.

Reply:
Yes, it is needed (but took us a while to figure out yesterday): Consider the case where the value we are considering is the loop head, at which widening is applied: Thus the old value we find in rho is the widened value. However, for a subsequent narrowing iteration we do not want to use the value AFTER widening, but we want to use the last value that the RHS returned, such that we can then benefit in the narrowing. This is evidenced e.g. by test 44/44. (Your `eq-reuse` in `solve` sometimes fixes it, but to work generally we need `oldv`).

Reply:
Ah, yes, that's right! I think in #364 I also had to have a special check that forbid aborting when switching from `Widen` to `Narrow` using `~abortable:false`. I think at the time, I didn't fully understand why, but your explanation makes sense.

I guess it's probably easier to explicitly keep `oldv` than trick around it and understand the exact combination of conditions when it cannot be used.